### ISSUE TITLE
Infinite while in case of wrong exception

### ISSUE THREAD

**Issue by gdoumenc** [ipython/ipython#9119]
In case a code generate an exception with this same exception as cause, the shell never render the exception and stay in loop.

To resolve 'partially' that, we can (al least) just add a simple one step test :
```python
(line 1019)
while evalue:
  formatted_exceptions += self.format_exception_as_a_whole(etype, evalue, etb, lines_of_context,
                                                                  chained_exceptions_tb_offset)
  exception = self.get_parts_of_chained_exception(evalue)

(----v)
  if exception and not exception[1] is evalue:
(----^)
```

**Comment by takluyver** [comment]
Good catch! To go one step better, we could keep a set of ids, and bail out if we reach an id we've already seen:

```python
chained_exc_ids = set()
while evalue:
    chained_exc_ids.add(id(evalue))

    ...
    exception = self.get_parts_of_chained_exception(evalue)
    if exception and id(exception[1]) not in chained_exc_ids:
        ...
```

(something like that, anyway)

Do you want to make a pull request for either of these methods?

**Comment by gdoumenc** [comment]
Your solution is better so I let you do your pull request.. ;)

FYI : I got this issue with pyodbc/sqlalchemy so not unknown package.... :(

Regards

Le 09/01/2016 17:27, Thomas Kluyver a écrit :
>
> Good catch! To go one step better, we could keep a set of ids, and 
> bail out if we reach an id we've already seen:
>
> chained_exc_ids=  set()
> while  evalue:
>      chained_exc_ids.add(id(evalue))
>
>      ...
>      exception=  self.get_parts_of_chained_exception(evalue)
>      if  exceptionand  id(exception[1])not  in  chained_exc_ids:
>          ...
>
> (something like that, anyway)
>
> Do you want to make a pull request for either of these methods?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/ipython/ipython/issues/9119#issuecomment-170257727>.
>

-- 
Guillaume DOUMENC
*FPR / CoWorks*
Your Internet Enabler
http://www.fpr-coworks.com

**Comment by takluyver** [comment]
You're still welcome to do a pull request if you're interested :-)

Have you also filed bugs with those packages? I think there's something wrong if an exception ends up being its own cause.

**Comment by takluyver** [comment]
Closed by #9121
